File: crates/iyon/src/tui/theme.rs
use iyon_api::ReasoningLevel;
use iyon_tui::{
    ColorSpec, StyleSelector, StyleSpec, StyleStateKey, StyleStateValue, TextSelector, Theme,
};

use super::transcript::thinking_tag;

pub(crate) const AGENT_EFFORT: StyleStateKey = StyleStateKey::from_static("iyon.agent.effort");
pub(crate) const EFFORT_NONE: StyleStateValue = StyleStateValue::from_static("none");
pub(crate) const EFFORT_MINIMAL: StyleStateValue = StyleStateValue::from_static("minimal");
pub(crate) const EFFORT_LOW: StyleStateValue = StyleStateValue::from_static("low");
pub(crate) const EFFORT_MEDIUM: StyleStateValue = StyleStateValue::from_static("medium");
pub(crate) const EFFORT_HIGH: StyleStateValue = StyleStateValue::from_static("high");
pub(crate) const EFFORT_XHIGH: StyleStateValue = StyleStateValue::from_static("xhigh");
pub(crate) const EFFORT_MAX: StyleStateValue = StyleStateValue::from_static("max");

pub(crate) fn effort_style_value(level: ReasoningLevel) -> StyleStateValue {
    match level {
        ReasoningLevel::None => EFFORT_NONE.clone(),
        ReasoningLevel::Minimal => EFFORT_MINIMAL.clone(),
        ReasoningLevel::Low => EFFORT_LOW.clone(),
        ReasoningLevel::Medium => EFFORT_MEDIUM.clone(),
        ReasoningLevel::High => EFFORT_HIGH.clone(),
        ReasoningLevel::XHigh => EFFORT_XHIGH.clone(),
        ReasoningLevel::Max => EFFORT_MAX.clone(),
    }
}

pub(crate) fn iyon_theme() -> Theme {
    Theme::new()
        .with_color(
            "surface.user",
            iyon_tui::ThemeColor::Rgb {
                r: 45,
                g: 55,
                b: 72,
            },
        )
        .with_color(
            "text.muted",
            iyon_tui::ThemeColor::Rgb {
                r: 113,
                g: 128,
                b: 150,
            },
        )
        .with_color(
            "surface.default",
            iyon_tui::ThemeColor::Rgb {
                r: 113,
                g: 128,
                b: 150,
            },
        )
        .with_color(
            "tool.running",
            iyon_tui::ThemeColor::Rgb {
                r: 160,
                g: 174,
                b: 192,
            },
        )
        .with_color(
            "tool.finished",
            iyon_tui::ThemeColor::Rgb {
                r: 104,
                g: 211,
                b: 145,
            },
        )
        .with_color(
            "tool.error",
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Red),
        )
        .with_color(
            "text.error",
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Red),
        )
        .with_color(
            "text.warning",
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Yellow),
        )
        .with_color(
            "text.heading",
            iyon_tui::ThemeColor::Rgb {
                r: 255,
                g: 196,
                b: 87,
            },
        )
        .with_color(
            "text.code",
            iyon_tui::ThemeColor::Rgb {
                r: 120,
                g: 200,
                b: 210,
            },
        )
        .with_color(
            "truncation_footer",
            iyon_tui::ThemeColor::Rgb {
                r: 120,
                g: 122,
                b: 132,
            },
        )
        .with_color(
            "input.border",
            iyon_tui::ThemeColor::Rgb {
                r: 173,
                g: 216,
                b: 230,
            },
        )
        .with_color_variant(
            "input.border",
            StyleSelector::state(AGENT_EFFORT.clone(), EFFORT_LOW.clone()),
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Green),
        )
        .with_color_variant(
            "input.border",
            StyleSelector::state(AGENT_EFFORT.clone(), EFFORT_MEDIUM.clone()),
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Yellow),
        )
        .with_color_variant(
            "input.border",
            StyleSelector::state(AGENT_EFFORT.clone(), EFFORT_HIGH.clone()),
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Magenta),
        )
        .with_color_variant(
            "input.border",
            StyleSelector::state(AGENT_EFFORT.clone(), EFFORT_LOW.clone()).and_focused(),
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::LightGreen),
        )
        .with_color_variant(
            "input.border",
            StyleSelector::state(AGENT_EFFORT.clone(), EFFORT_HIGH.clone()).and_focused(),
            iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::LightMagenta),
        )
        .with_style(
            "tool.running",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("tool.running")),
        )
        .with_style(
            "tool.finished",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("tool.finished")),
        )
        .with_style(
            "tool.error",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("tool.error")),
        )
        .with_style(
            "text.muted",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("text.muted")),
        )
        .with_style(
            "text.error",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("text.error")),
        )
        .with_style(
            "text.warning",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("text.warning")),
        )
        .with_text_style(
            TextSelector::heading(),
            StyleSpec::new().foreground(ColorSpec::theme("text.heading")),
        )
        .with_text_style(
            TextSelector::inline_code(),
            StyleSpec::new().foreground(ColorSpec::theme("text.code")),
        )
        .with_text_style(
            TextSelector::code_block(),
            StyleSpec::new().foreground(ColorSpec::theme("text.code")),
        )
        .with_text_style(
            TextSelector::annotation(&thinking_tag()),
            StyleSpec::new()
                .foreground(ColorSpec::theme("text.muted"))
                .italic(),
        )
        .with_style(
            "truncation_footer",
            iyon_tui::StyleSpec::new().foreground(iyon_tui::ColorSpec::theme("truncation_footer")),
        )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn palette_preserves_exact_named_and_rgb_values() {
        let theme = iyon_theme();
        assert_eq!(
            theme.color("surface.user"),
            Some(iyon_tui::ThemeColor::Rgb {
                r: 45,
                g: 55,
                b: 72
            })
        );
        assert_eq!(
            theme.color("tool.finished"),
            Some(iyon_tui::ThemeColor::Rgb {
                r: 104,
                g: 211,
                b: 145
            })
        );
        assert_eq!(
            theme.color("tool.error"),
            Some(iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Red))
        );
        assert_eq!(
            theme.color("text.warning"),
            Some(iyon_tui::ThemeColor::Named(iyon_tui::AnsiColor::Yellow))
        );
        assert_eq!(
            theme.color("text.heading"),
            Some(iyon_tui::ThemeColor::Rgb {
                r: 255,
                g: 196,
                b: 87
            })
        );
        assert_eq!(
            theme.color("text.code"),
            Some(iyon_tui::ThemeColor::Rgb {
                r: 120,
                g: 200,
                b: 210
            })
        );
        assert!(theme.style("markdown.header").is_none());
        assert!(theme.style("markdown.code").is_none());
    }

    #[test]
    fn assistant_text_styles_come_from_theme_not_the_renderer() {
        use crate::tui::transcript::pipeline::assistant_renderer;
        use iyon_tui::text::{CodeBlock, Inline, InlineContent, LanguageId, LiteralText, Renderer};
        use iyon_tui::{Block, HeadingLevel, MarkdownOptions, MarkdownProjector};

        let renderer = assistant_renderer();
        assert_eq!(
            renderer.policy().soft_break(),
            iyon_tui::SoftBreakPolicy::LineBreak
        );
        assert_eq!(
            MarkdownProjector::default().options(),
            MarkdownOptions::commonmark()
        );

        let theme = iyon_theme();
        let heading = renderer.render(&Block::heading(HeadingLevel::H1, "Title"));
        let heading_style = iyon_tui::testing::style_at_text(&heading, 40, &theme, "Title");
        assert_eq!(heading_style.fg_rgb, Some((255, 196, 87)));
        assert!(heading_style.bold);
        assert!(heading_style.underline);

        let code = renderer.render(&Block::paragraph(InlineContent::new([Inline::text(
            "code",
        )
        .code()])));
        assert_eq!(
            iyon_tui::testing::style_at_text(&code, 40, &theme, "code").fg_rgb,
            Some((120, 200, 210))
        );

        let rust = LanguageId::new("rust").unwrap();
        let block = renderer.render(&Block::code(CodeBlock::new(
            Some(rust),
            Some("rust"),
            LiteralText::from("fn"),
        )));
        assert_eq!(
            iyon_tui::testing::style_at_text(&block, 40, &theme, "fn").fg_rgb,
            Some((120, 200, 210))
        );

        let think = iyon_tui::text::TextRun::from("think")
            .map_annotations(|annotations| annotations.with_tag(crate::transcript::thinking_tag()));
        let thinking_view =
            renderer.render(&Block::paragraph(InlineContent::new([Inline::text(think)])));
        let muted = iyon_tui::testing::style_at_text(&thinking_view, 40, &theme, "think");
        assert_eq!(muted.fg_rgb, Some((113, 128, 150)));
        assert!(muted.italic);

        let think_code = iyon_tui::text::TextRun::from("fnx")
            .map_annotations(|annotations| annotations.with_tag(crate::transcript::thinking_tag()));
        let thinking_code = renderer.render(&Block::paragraph(InlineContent::new([Inline::text(
            think_code,
        )
        .code()])));
        let overlay = iyon_tui::testing::style_at_text(&thinking_code, 40, &theme, "fnx");
        assert_eq!(overlay.fg_rgb, Some((113, 128, 150)));
        assert!(overlay.italic);
    }
}

File: plugins/app/iyon/src/theme.ts
import { Style, StyleSelector, TextSelector, Theme, themeColor } from "@iyon/tui";
import type {
  AnsiColor,
  StyleSpec,
  Theme as RuntimeTheme,
  ThemeColor,
  ThemeColorReference,
} from "@iyon/tui";

export type IyonTheme = RuntimeTheme & {
  readonly composer: StyleSpec;
  readonly footer: StyleSpec;
  readonly muted: StyleSpec;
  readonly inputBorder: ThemeColorReference;
  readonly mutedColor: ThemeColorReference;
  readonly toolFinishedColor: ThemeColorReference;
};

function rgb(r: number, g: number, b: number): ThemeColor {
  return { type: "rgb", r, g, b };
}

function ansi(value: AnsiColor): ThemeColor {
  return { type: "named", value };
}

function effort(value: string): StyleSelector {
  return StyleSelector.state("iyon.agent.effort", value);
}

function focusedEffort(value: string): StyleSelector {
  return StyleSelector.focused().andState("iyon.agent.effort", value);
}

export function createIyonTheme(effortOverride?: string): IyonTheme {
  const currentEffort = effortOverride ?? "medium";
  const accentColor =
    currentEffort === "none" ? ansi("white")
    : currentEffort === "minimal" ? ansi("lightCyan")
    : currentEffort === "low" ? ansi("green")
    : currentEffort === "medium" ? rgb(255, 196, 87)
    : currentEffort === "high" ? ansi("magenta")
    : currentEffort === "xhigh" ? ansi("lightMagenta")
    : currentEffort === "max" ? ansi("lightRed")
    : rgb(255, 196, 87);
  // text.code stays constant — only text.heading changes with effort
  const codeColor = rgb(120, 200, 210);
  const diffHeaderColor = rgb(255, 196, 87);

  // Border color for each effort level (used as base + for frozen messages).
  const inputBorderColor =
    currentEffort === "none" ? ansi("white")
    : currentEffort === "minimal" ? ansi("lightCyan")
    : currentEffort === "low" ? ansi("green")
    : currentEffort === "medium" ? ansi("yellow")
    : currentEffort === "high" ? ansi("magenta")
    : currentEffort === "xhigh" ? ansi("lightMagenta")
    : currentEffort === "max" ? ansi("lightRed")
    : rgb(173, 216, 230);
  const theme = Theme.new()
    .withColor("surface.user", rgb(45, 55, 72))
    .withColor("text.muted", rgb(113, 128, 150))
    .withColor("surface.default", rgb(113, 128, 150))
    .withColor("tool.running", rgb(160, 174, 192))
    .withColor("tool.finished", rgb(104, 211, 145))
    .withColor("tool.error", ansi("red"))
    .withColor("text.error", ansi("red"))
    .withColor("text.warning", ansi("yellow"))
    .withColor("text.heading", accentColor)
    .withColor("text.code", codeColor)
    .withColor("diff.addition", rgb(104, 211, 145))
    .withColor("diff.deletion", ansi("red"))
    .withColor("diff.header", diffHeaderColor)
    .withColor("diff.context", rgb(113, 128, 150))
    .withColor("diff.meta", rgb(113, 128, 150))
    .withColor("truncation_footer", rgb(120, 122, 132))
    .withColor("input.border", inputBorderColor)
    .withColorVariant("input.border", effort("none"), ansi("white"))
    .withColorVariant("input.border", effort("minimal"), ansi("lightCyan"))
    .withColorVariant("input.border", effort("low"), ansi("green"))
    .withColorVariant("input.border", effort("medium"), ansi("yellow"))
    .withColorVariant("input.border", effort("high"), ansi("magenta"))
    .withColorVariant("input.border", effort("xhigh"), ansi("lightMagenta"))
    .withColorVariant("input.border", effort("max"), ansi("lightRed"))
    .withColorVariant("input.border", focusedEffort("none"), ansi("white"))
    .withColorVariant("input.border", focusedEffort("minimal"), ansi("lightCyan"))
    .withColorVariant("input.border", focusedEffort("low"), ansi("lightGreen"))
    .withColorVariant("input.border", focusedEffort("medium"), ansi("lightYellow"))
    .withColorVariant("input.border", focusedEffort("high"), ansi("lightMagenta"))
    .withColorVariant("input.border", focusedEffort("xhigh"), ansi("lightMagenta"))
    .withColorVariant("input.border", focusedEffort("max"), ansi("lightRed"))
    .withStyle("tool.running", Style.new().foreground(themeColor("tool.running")))
    .withStyle("tool.finished", Style.new().foreground(themeColor("tool.finished")))
    .withStyle("tool.error", Style.new().foreground(themeColor("tool.error")))
    .withStyle("text.muted", Style.new().foreground(themeColor("text.muted")))
    .withStyle("text.error", Style.new().foreground(themeColor("text.error")))
    .withStyle("text.warning", Style.new().foreground(themeColor("text.warning")))
    .withStyle("diff.addition", Style.new().foreground(themeColor("diff.addition")))
    .withStyle("diff.deletion", Style.new().foreground(themeColor("diff.deletion")))
    .withStyle("diff.header", Style.new().foreground(themeColor("diff.header")))
    .withStyle("diff.context", Style.new().foreground(themeColor("diff.context")))
    .withStyle("diff.meta", Style.new().foreground(themeColor("diff.meta")))
    .withTextStyle(TextSelector.heading(), Style.new().foreground(themeColor("text.heading")))
    .withTextStyle(TextSelector.inlineCode(), Style.new().foreground(themeColor("text.code")))
    .withTextStyle(TextSelector.codeBlock(), Style.new().foreground(themeColor("text.code")))
    .withTextStyle(TextSelector.part("codeLabel"), Style.new().foreground(themeColor("text.muted")).dim())
    .withTextStyle(TextSelector.part("quoteMarker"), Style.new().foreground(themeColor("text.muted")))
    .withTextStyle(TextSelector.part("listMarker"), Style.new().foreground(themeColor("text.muted")))
    .withTextStyle(TextSelector.part("taskMarker"), Style.new().foreground(themeColor("text.muted")))
    .withTextStyle(TextSelector.part("thematicRule"), Style.new().foreground(themeColor("text.muted")))
    .withTextStyle(TextSelector.annotation("app", "thinking"), Style.new().foreground(themeColor("text.muted")).italic());

  return Object.assign(theme, {
    composer: Style.new(),
    footer: Style.new().dim(),
    muted: Style.new().dim().foreground(themeColor("text.muted")),
    inputBorder: themeColor("input.border"),
    mutedColor: themeColor("text.muted"),
    toolFinishedColor: themeColor("tool.finished"),
  });
}
